### PR TITLE
Allow dispatch region formation from scf.foreach_thread with proper tensor semantics.

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This revision introduces a direct rewrite of an scf::ForeachThreadOp to a DispatchWorkgroupsOp which does not require a 
2-step creation + legalization and generally simplifies the implementation.
This is possible thanks to the scf::ForeachThreadOp abstraction that closely matches a DispatchWorkgroupsOp.
    
Opportunities for API / builder / verification improvements of the DispatchWorkgroupsOp are exposed by this implementation.
    
Also adds an transform.iree.foreach_thread_to_flow transform to the transform dialect extensions.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
Are all the commits from Ben part of this CL? Or is this some rebasing issue?

**Comment by nicolasvasilache** [comment]
it's a rebasing thing, sorry I am not sure why they show up here, updating.

**Comment by nicolasvasilache** [comment]
done, sorry for the noise

**Comment by hanhanW** [comment]
The change is landed

**Comment by hanhanW** [comment]
comment on wrong PR.. sorry

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/TilingToForeachThreadOp.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/OperationSupport.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::iree_compiler::IREE::LinalgExt;
+
+static FailureOr<TilingResult> tileToForeachOp(PatternRewriter &rewriter,
```
Comment (matthias-springer) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/TilingToForeachThreadOp.cpp:24]: This could be a separate PR

Comment (nicolasvasilache) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/TilingToForeachThreadOp.cpp:24]: see #9288

File: compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp [hunk 2]
```diff
@@ -41,6 +46,329 @@ void mlir::iree_compiler::registerLinalgTransformDialectExtension(
...
+    PatternRewriter &rewriter, scf::PerformConcurrentlyOp performConcurrentlyOp,
+    Block &block, ValueRange resultTensorOperands,
+    ValueRange resultTensorsDynamicDims, BlockAndValueMapping tensorToFlowBvm) {
+  Location loc = performConcurrentlyOp.getLoc();
+  int64_t resultIndex = 0;
+  for (const Operation &yieldingOp :
+       llvm::make_early_inc_range(performConcurrentlyOp.yieldingOps())) {
+    auto parallelInsertOp = cast<scf::ParallelInsertSliceOp>(&yieldingOp);
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(block.getTerminator());
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp:75]: this wants to insert **before** the terminator, what you suggest would insert after it.

File: compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp [hunk 3]
```diff
@@ -41,6 +46,315 @@ void mlir::iree_compiler::registerLinalgTransformDialectExtension(
...
+  block->addArguments(
+      SmallVector<Type>(allTensorDynamicDims.size(), rewriter.getIndexType()),
+      SmallVector<Location>(allTensorDynamicDims.size(), loc));
+  SmallVector<Value> allOperands = nonDimOperands;
+  llvm::append_range(allOperands, allTensorDynamicDims);
+  assert(block->getNumArguments() == allOperands.size() &&
+         "Expected as many bbArgs as operands");
+
+  // Step 6. Move the body of inParallel to the dispatchOp.
+  block->getOperations().splice(
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp:291]: In principle yes, however in this case, it does require knowledge of the order of the bbArgs to properly order the bbArgReplacements which leaks implementation details about the op build method. In the future we will want better helpers on the op side, which is why we have the `TODO`.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Could we move the pass that creates the dispatches into a separate file? It's better not to bloat the file that creates dispatches today.

**Comment by MaheshRavishankar** [comment]
Looks mostly reasonable. Left a few minor comments. Would be good for @benvanik to review as well. There is some wonkiness with respect to reading the transformation from an MLIR file (which is OK for now), but as the owner here, might be good to get sign off from Ben about it.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  if (!memoryBuffer) {
+    llvm::errs() << errorMessage << "\n";
+    return signalPassFailure();
+  }
+
+  // Tell sourceMgr about this buffer, the parser will pick it up.
+  llvm::SourceMgr sourceMgr;
+  sourceMgr.AddNewSourceBuffer(std::move(memoryBuffer), llvm::SMLoc());
+  OwningOpRef<ModuleOp> transformModule(
+      parseSourceFile<ModuleOp>(sourceMgr, &getContext()));
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:89]: Is this thread-safe? (this pass will be called concurrently)

We should at least memoize it - this isn't a cheap operation and when using the transform dialect reusing the transformModule is going to be important anywhere it's used.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td [hunk 5]
```diff
@@ -56,6 +56,12 @@ def DispatchLinalgOnTensors :
   let constructor = "mlir::iree_compiler::IREE::Flow::createDispatchLinalgOnTensorsPass()";
 }
 
+def DispatcWithTransformDialect:
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td:59]:
```suggestion
def DispatchWithTransformDialect :
```

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_transform_dialect.mlir [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+func.func @tile_matmul_alone(%arg0 : tensor<?x?xf32>, %arg1 : tensor<?x?xf32>,
+             %arg2 : tensor<?x?xf32>) -> tensor<?x?xf32> {
+  %1 = linalg.matmul ins(%arg0, %arg1 : tensor<?x?xf32>, tensor<?x?xf32>)
+    outs(%arg2 : tensor<?x?xf32>) -> tensor<?x?xf32>
+  return %1 : tensor<?x?xf32>
+}
+
+//      CHECK: func.func @tile_matmul_alone
+//      CHECK:   flow.dispatch.workgroups
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_transform_dialect.mlir:13]: style: missing newline at end of file (maybe check your editor settings?)

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 The IREE Authors
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:1]: style: current year on new files
```suggestion
// Copyright 2022 The IREE Authors
```

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Transform/IR/TransformDialect.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/FileUtilities.h"
+#include "mlir/Support/LLVM.h"
+
+static llvm::cl::opt<std::string> clTransformFileName(
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:27]: It'd be good to make this a pass option instead as then reproducers/etc work with it and it's scoped to the pass.
Then the -iree-flow- flag for enabling the pass could change to a string that is passed in to set the option and simplify the code paths (since it's not valid to use the flag without a file set and doesn't make sense to accept the filename param unless the flag is set).

```
--iree-flow-dispatch-use-transform-dialect --dispatch-transform-dialect-file-name=%p/transform_dialect_dispatch_spec.mlir 
```
->
```
--iree-flow-dispatch-use-transform-dialect=%p/transform_dialect_dispatch_spec.mlir 
```

**Comment by benvanik** [comment]
This layers in nicely!

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp (hunk 8 above)
Comment (stellaraccident) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:27]: I've got some ideas on how this can come together better for such reproducers and be hermetic. Early in the pipeline, we probably want to read the file and inline it into the module, then using the module bound transformations from then on...

Comment (stellaraccident) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:27]: (for later)

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.h [hunk 9]
```diff
@@ -136,6 +136,11 @@ std::unique_ptr<Pass> createVerifyInputLegalityPass();
 std::unique_ptr<InterfacePass<mlir::FunctionOpInterface>>
 createDispatchLinalgOnTensorsPass();
 
+// Pass to perform dispatch of Linalg on tensor ops by using the transform
+// dialect. Dispatch regions are created as specified by the tr.
```
Comment (stellaraccident) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.h:140]: Don't abbreviate "tr" at the end.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp [hunk 10]
```diff
@@ -38,7 +38,7 @@
 using namespace mlir;
 
 static llvm::cl::opt<std::string> clTransformFileName(
-    "linalg-transform-file-name",
+    "codegen-transform-dialect-file-name",
```
Comment (stellaraccident) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp:41]: Fine for now, but we're going to want to fix/generalize the way this is configured before using by default.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp (hunk 4 above)
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:89]: Good point!
Note that this is definitely not ready for prime time but only to allow connecting pieces for now.
Do you have a suggestion as to where that memoization should occur?
I can lift this into the initializer for now but this leads to a bad use of a shared_ptr due to various fun things around MLIR passes.
Please also see my batch reply/question for more discussion.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp (hunk 8 above)
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:27]: `we probably want to read the file and inline it into the module, then using the module bound transformations from then on` -> this was the first thing I tried a while back but it really wasn't clear how to transport this module properly in the IR the the proper place an through IREE transformations and IR massaging. The file was the only sane thing I could come up with at the time. Maybe it's time to revisit but it will prob require deeper design.

Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:27]: capturing some of this in my batch reply and resolving this instance.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp (hunk 10 above)
Comment (nicolasvasilache) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp:41]: same here re batch reply and larger design Qs

**Comment by nicolasvasilache** [comment]
Thanks for the reviews, responding in batch to the following comments:

> There is some wonkiness with respect to reading the transformation from an MLIR file
> Is this thread-safe? We should at least memoize it - this isn't a cheap operation and when using the transform dialect reusing the transformModule is going to be important anywhere it's used.
> I've got some ideas on how this can come together better for such reproducers and be hermetic. Early in the pipeline, we probably want to read the file and inline it into the module, then using the module bound transformations from then on...
> Fine for now, but we're going to want to fix/generalize the way this is configured before using by default.

My first try was to embed transform IR in the original input IR but it really wasn't clear how to transport this module properly in the IR to the proper place. The file was the only sane thing I could come up with at the time. Maybe it's time to revisit but it will prob require deeper design.

There are a few colliding issues and the impl evolved to this in this current PR:
```
struct DispatchWithTransformDialect ... {
  ...
  DispatchWithTransformDialect(const DispatchWithTransformDialect &pass) {
    transformFileName = pass.transformFileName;
    // TODO: if we really don't like shared_ptr, we could also clone the
    // transformModule here.
    sharedTransformModule = pass.sharedTransformModule;
  }
  LogicalResult initialize(MLIRContext *context) override {
    ...
    sharedTransformModule = parseSourceFile<ModuleOp>(sourceMgr, context);
  }
  ...

  // The parsed transform module to be used for creating dispatches.
  // TODO: Figure a better way to build a transform module and transport it in
  // the proper places in the IR as it is transformed by IREE so that it is
  // available with better ownership semantics.
  // Note: we wrap the OwningOpRef to get the desired destruction mechanism.
  // Note: shared_ptr is not great but we know the sharedTransformModule is
  // readonly.
  // Alternatives comprise:
  //   1. no shared_ptr but copying the module with every pass clone that the
  //      OpPassManager decides to perform.
  //   2. lifting ownership of the parsed transform module higher up in the 
  //      IREE stack. This may be only shift the problem as we have passes 
  //      building pass managers in IREE.
  //   3. build better support to embed the transformation module in the 
  //      input IR and transport it to the place of use in IREE. This is deemed
  //      too intrusive atm.
  //   4. (future) config/resources mechanism that is being proposed in core?
  std::shared_ptr<OwningOpRef<ModuleOp>> sharedTransformModule;
  llvm::StringRef transformFileName;
}
```

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp (hunk 8 above)
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp]: @benvanik it works for the e2e test but I am stuck with passing the proper pass options to `compiler/src/iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_transform_dialect.mlir` to make an iree-opt unit test run.

Before I had:
```
// RUN: iree-opt --dispatch-transform-dialect-file-name=%p/transform_dialect_dispatch_spec.mlir \
// RUN:   --pass-pipeline="func.func(iree-flow-dispatch-linalg-on-tensors-transform-dialect-pass), cse, canonicalize, cse" 
```

Now I need:
```
// RUN: iree-opt --iree-flow-dispatch-use-transform-dialect=%p/transform_dialect_dispatch_spec.mlir \
// RUN:   --pass-pipeline="???" 
```

Could you help find the right incantation?

**Comment by stellaraccident** [comment]
> Thanks for the reviews, responding in batch to the following comments:
> 
> > There is some wonkiness with respect to reading the transformation from an MLIR file
> > Is this thread-safe? We should at least memoize it - this isn't a cheap operation and when using the transform dialect reusing the transformModule is going to be important anywhere it's used.
> > I've got some ideas on how this can come together better for such reproducers and be hermetic. Early in the pipeline, we probably want to read the file and inline it into the module, then using the module bound transformations from then on...
> > Fine for now, but we're going to want to fix/generalize the way this is configured before using by default.
> 
> My first try was to embed transform IR in the original input IR but it really wasn't clear how to transport this module properly in the IR to the proper place. The file was the only sane thing I could come up with at the time. Maybe it's time to revisit but it will prob require deeper design.
> 
> There are a few colliding issues and the impl evolved to this in this current PR:
> 
> ```
> struct DispatchWithTransformDialect ... {
>   ...
>   DispatchWithTransformDialect(const DispatchWithTransformDialect &pass) {
>     transformFileName = pass.transformFileName;
>     // TODO: if we really don't like shared_ptr, we could also clone the
>     // transformModule here.
>     sharedTransformModule = pass.sharedTransformModule;
>   }
>   LogicalResult initialize(MLIRContext *context) override {
>     ...
>     sharedTransformModule = parseSourceFile<ModuleOp>(sourceMgr, context);
>   }
>   ...
> 
>   // The parsed transform module to be used for creating dispatches.
>   // TODO: Figure a better way to build a transform module and transport it in
>   // the proper places in the IR as it is transformed by IREE so that it is
>   // available with better ownership semantics.
>   // Note: we wrap the OwningOpRef to get the desired destruction mechanism.
>   // Note: shared_ptr is not great but we know the sharedTransformModule is
>   // readonly.
>   // Alternatives comprise:
>   //   1. no shared_ptr but copying the module with every pass clone that the
>   //      OpPassManager decides to perform.
>   //   2. lifting ownership of the parsed transform module higher up in the 
>   //      IREE stack. This may be only shift the problem as we have passes 
>   //      building pass managers in IREE.
>   //   3. build better support to embed the transformation module in the 
>   //      input IR and transport it to the place of use in IREE. This is deemed
>   //      too intrusive atm.
>   //   4. (future) config/resources mechanism that is being proposed in core?
>   std::shared_ptr<OwningOpRef<ModuleOp>> sharedTransformModule;
>   llvm::StringRef transformFileName;
> }
> ```

I haven't fully considered all of the options yet. But I'm thinking that a flag like you originally had it always exists and we do something at the top of the compiler to read the file and stitch it into the module being compiled. If we set that up right, we could also support something else inlining the transforms prior to compilation, which will be important for search, et al.

I would just leave this is a global flag or pass option for now (whatever works for testing) and make sure that all access to the transform module is though a single helper that we can make smarter.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td [hunk 11]
```diff
@@ -56,6 +56,12 @@ def DispatchLinalgOnTensors :
   let constructor = "mlir::iree_compiler::IREE::Flow::createDispatchLinalgOnTensorsPass()";
 }
 
+def DispatchWithTransformDialect :
+    InterfacePass<"iree-flow-dispatch-linalg-on-tensors-transform-dialect-pass", "mlir::FunctionOpInterface"> {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td:60]: we try to keep pass names matching the C++/file names for clarity;
with a filename of DispatchWithTransformDialect this should be `iree-flow-dispatch-with-transform-dialect` (no `-pass`)

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp (hunk 8 above)
Comment (stellaraccident) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp]: I think that's the high order bit for me: of going beyond testing/experimentation, this needs more thought/discussion and I'm fine with something simple at this stage.

If river's resources stuff works out, we may have a path to just bundle it for reproducers without messing with everything else. But that is future.

**Comment by benvanik** [comment]
I think as long as we can check that it's not going to cause crashes the first time there are two functions in a module it's fine as it is. IIRC @bjacob did some quashing of threading bugs and may have some guidance/references for this kind of stuff and how to avoid the pitfalls.

**Comment by bjacob** [comment]
Happy to chat if there's any specific question here, otherwise I can only suggest giving it a good shake with TSan :-)  We just updated the docs here:

https://github.com/google/iree/blob/main/docs/developers/developing_iree/sanitizers.md#tsan-threadsanitizer

We're in the process of setting up TSAN CI bot and might have it any time this week, too (#9371).

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp (hunk 8 above)
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp]: After more spelunking, passing shared IR to the transform dialect and PDL is not a viable solution. The underlying issue is that any IR manipulation in LLVM that performs a move of an op (e.g. splice) modifies **surrounding** IR (i.e. the links of the op into the surrounding block) in a non-threadsafe way.

In the longer term, extracting transform IR from a larger database and attaching it to the place where it is applied in the IR is likely the way to go.

For now, parsing is done at initialization time and cloning at `runOperation` time, as this would be a good place to perform policy filtering based on actual IR. 
Alternatively, we may also want to just clone the transform IR on each thread; this will not have filtering capabilities but OTOH such filtering will likely be its own orthogonal mechanism.

Please lmk if you have a strong preference one way of another, in any case this will disappear in some future and other mechanisms will be needed to select transform IR.

**Comment by nicolasvasilache** [comment]
@benvanik @bjacob thanks, there was indeed a non-trivial behavior related to clone + move under MT execution.

**Review by benvanik — APPROVED** [review]

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Github is strange sometimes. I left some comments from the last time around but somehow they werent published. Pushing those comments now. Leaving it as request changes for now, Ill stamp in the morning

**Review by MaheshRavishankar — APPROVED** [review]
I have some comments about some transformations that are better suited for putting in Flow. But overall this is fine. Stamped

File: compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp [hunk 12]
```diff
@@ -41,6 +46,315 @@ void mlir::iree_compiler::registerLinalgTransformDialectExtension(
...
+      loc,
+      /*workgroupCount=*/workgroupCount,
+      /*resultTypes=*/foreachThreadOp.getResultTypes(),
+      /*resultDims=*/resultTensorsDynamicDims.getArrayRef(),
+      /*operands=*/nonDimOperands,
+      /*operandDims=*/allTensorDynamicDims,
+      /*tiedOperands=*/llvm::to_vector<4>(tiedOperandsSequence));
+  // clang-format on
+
+  // Step 5. Fixup dispatchOp bbArgs and terminator.
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp:250]: Yeah I saw that the helper function that infers stuff is clever and allows arbitrary mixing of tensors and dims as long as the dims come after the tensors.
In the op itself, it would be nice to have some verification and at least document an expected way we would want to read them.

File: compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp [hunk 13]
```diff
@@ -41,6 +46,315 @@ void mlir::iree_compiler::registerLinalgTransformDialectExtension(
...
+///   - Step 7: Set up bvm for RAUWIf. In particular, tensor operands become
+///     flow dispatch tensor bbArgs and need to be
+///     flow.dispatch.tensor.load'ed.
+///   - Step 8: Plug dispatch workgroup id and count values into the bvm.
+///   - Step 9. Rewrite tensor::ExtractSlice and ParallelInsert ops to the
+///     relevant Flow DispatchTensorLoad/Store version.
+///   - Step 10: Perform RAUWIf.
+///   - Step 11: Drop the terminator and replace foreachThreadOp.
+// TODO: n-D ForeachThreadOp
+struct ForeachThreadToFlowDispatchWorkgroupsRewriter
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp:150]: This is to avoid the complexity of setting up another set of files for the purpose of registering a single op in the transform dialect. I can add those as a follow up.

**Comment by bjacob** [comment]
I just checked, we are still TSan green after this PR. congrats :grin:
